coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala

```diff
@@ -39,595 +23,32 @@ object UnavailablePromQlQueryParams extends TsdbQueryParams
 class QueryEngine(dsRef: DatasetRef,
                   schemas: Schemas,
                   shardMapperFunc: => ShardMapper,
+//                  downsampleMapperFunc: => ShardMapper,
                   failureProvider: FailureProvider,
                   spreadProvider: SpreadProvider = StaticSpreadProvider(),
-                  queryEngineConfig: Config = ConfigFactory.empty())
-                   extends StrictLogging {
-  /**
-    * Intermediate Plan Result includes the exec plan(s) along with any state to be passed up the
-    * plan building call tree during query planning.
-    *
-    * Not for runtime use.
-    */
-  private case class PlanResult(plans: Seq[ExecPlan], needsStitch: Boolean = false)
+                  queryEngineConfig: Config = ConfigFactory.empty()) extends QueryPlanner with StrictLogging {
 
-  private val mdNoShardKeyFilterRequests = Kamon.counter("queryengine-metadata-no-shardkey-requests")
-  private val dsOptions = schemas.part.options
+  // Note the composition of query planners below using decorator pattern
+  val rawClusterPlanner = new SingleClusterPlanner(dsRef, schemas, spreadProvider, shardMapperFunc)
+  val downsampleClusterPlanner = new SingleClusterPlanner(dsRef, schemas, spreadProvider, shardMapperFunc)
+  val downsampleStitchPlanner = new DownsampleStitchPlanner(rawClusterPlanner, downsampleClusterPlanner)
+  // TODO haPlanner should later use downsampleStitchPlanner
+  val haPlanner = new HighAvailabilityPlanner(dsRef, rawClusterPlanner, failureProvider,
+                                              spreadProvider, queryEngineConfig)
+  //val multiPodPlanner = new MultiPodPlanner(podLocalityProvider, haPlanner)
 
   /**
     * This is the facade to trigger orchestration of the ExecPlan.
     * It sends the ExecPlan to the destination where it will be executed.
     */
-  def dispatchExecPlan(execPlan: ExecPlan)
-                      (implicit sched: Scheduler,
-                       timeout: FiniteDuration): Task[QueryResponse] = {
+  def dispatchExecPlan(execPlan: ExecPlan)(implicit sched: Scheduler, timeout: FiniteDuration): Task[QueryResponse] = {
```

Comment:
Perhaps this utility can go into QueryPlanner? Then QueryEngine goes away. It's really a way of gluing these planners together.

Reply:
Good idea.